plugins/toolkit-backend/src/service/controller/toolkit.controller.ts

```diff
@@ -0,0 +1,186 @@
+      const { logo, title, url, type } = req.body;
+      const toolkit = await client('toolkit')
+        .update({
+          title,
+          logo,
+          url,
+          type,
+        })
+        .where({ id: req.params.id });
+      if (toolkit) {
+        if (type === 'private') {
+          await client('toolkit_item')
+            .where({ toolkit: req.params.id })
+            .whereNot({ user: user.identity.userEntityRef })
+            .delete();
+        }
+        res.status(200).send('Updated');
+      } else {
+        res.status(400).send('Something went wrong');
+      }
+    } else {
+      const toolkit = await client
+        .insert(req.body)
+        .into('toolkit')
+        .returning('id');
+      if (toolkit.length) {
+        const { id } = toolkit[0];
+        await client
+          .insert({
+            toolkit: id,
+            isPrivate: req.body.type === 'private',
+            user: req.body.owner,
+          })
+          .into('toolkit_item');
+        res.status(200).send('Created');
+      } else {
+        res.status(400).send('Something went wrong');
+      }
+    }
+  } catch (error: any) {
+    res.status(500).send('Internal server error!');
+  }
+};
+
+export const addToolkit = async (req: any, res: Response) => {
+  try {
+    const client = req.dbClient as Knex<any, any[]>;
+    const { toolkits } = req.body;
+    const user = req.user.identity.userEntityRef;
+    const toolkitData: any[] = [];
+    const previousData = await client('toolkit_item')
+      .where({ user })
+      .select('toolkit');
+    const checkIsExists = await client('toolkit')
+      .whereIn('id', toolkits)
+      .where({ type: 'public' });
+    toolkits.forEach((_id: number) => {
+      const checkPrev = previousData.find(({ toolkit }) => toolkit === _id);
+      const checkExists = checkIsExists.find(({ id }) => id === _id);
+      if (!checkPrev && checkExists) {
+        toolkitData.push({
+          toolkit: checkExists.id,
+          isPrivate: false,
+          user: user,
+        });
+      }
+    });
+
+    if (toolkitData?.length) {
+      await client.insert([...toolkitData]).into('toolkit_item');
+      res.send('Toolkits added successfully');
+    } else {
+      res.send('No Toolkits founded or already exists');
+    }
+  } catch (error: any) {
+    res.status(500).send('Internal server error!');
+  }
+};
+
+export const myToolkits = async (req: any, res: Response) => {
+  try {
+    const client = req.dbClient as Knex<any, any[]>;
+    const user = req.user;
+    const userId = user.identity.userEntityRef;
+    const condition: any = {
+      'toolkit_item.user': userId,
+    };
+
+    const data = await client('toolkit_item')
+      .where(condition)
+      .innerJoin('toolkit', 'toolkit.id', 'toolkit_item.toolkit');
+    res.status(200).json(data);
+  } catch (error) {
+    res.status(500).send('Internal server error!');
+  }
+};
+
+export const toolkitById = async (req: any, res: Response) => {
```

Comment:
could we improve the typings of all these methods in this file, accepting/returning concrete data instead of express' requests and responses? Otherwise it is going to be hard to catch errors.

For example this method could become:
`function toolkitById(id: string): Toolkit`

It will help us catching up a lot of edge cases, as:

```typescript
const data = await client('toolkit').where({ id: req.params.id }).first();`
res.status(200).json(data);
```

in case no items are matched `data` will be undefined, which with the current flow will break the frontend

Reply:
Fixed